lib/paperclip/io_adapters/data_uri_adapter.rb

```diff
@@ -10,10 +12,74 @@ def initialize(target_uri)
     private
 
     def extract_target(uri)
-      data_uri_parts = uri.match(REGEXP) || []
-      StringIO.new(Base64.decode64(data_uri_parts[2] || ''))
+      @data_uri_parts = (uri.match(REGEXP) || []).to_a
+
+      @payload = get_payload_from_match
```

Comment:
There is a lot of code dealing with parsing the URI. It seems like extracting an object to represent the data URI and handle the parsing would be a good idea here. That would leave this class to handle the Adapter parts.